src/Interpreters/TemporaryDataOnDisk.cpp

```diff
@@ -334,10 +347,28 @@ Block TemporaryFileStream::read()
     {
         /// finalize earlier to release resources, do not wait for the destructor
         this->release();
+        in_reader.reset();
     }
     return block;
 }
 
+std::unique_ptr<InputReader> TemporaryFileStream::getReadStream()
+{
+    if (!isWriteFinished())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Writing has been not finished");
+
+    if (isEof())
+        return nullptr;
+
+    if (auto type = read_type.exchange(2); type == 1)
+    {
+        read_type.store(1);
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Different type of reading was requested earlier");
+    }
```

Comment:
Maybe we can just check that `in_reader == nullptr` which means that `TemporaryFileStream::read` wasn't called

Reply:
but it is not async safe, am i wrong?